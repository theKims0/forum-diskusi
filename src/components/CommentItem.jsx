import React from 'react';
import PropTypes from 'prop-types';
import postedAt from '../utils';

function CommentItem({
  content = '',
  createdAt = '',
  owner = {},
}) {
  return (
    <section className="thread-reply">

      <header className="footer-note">
        <img className="img-round" src={owner.avatar} alt={owner} />
        <div className="user-info">
          <p>
            @
            {owner.id}
          </p>
          <span>
            {owner.name}
          </span>
        </div>
      </header>
      <article>
        <p>{content}</p>
      </article>
      <footer className="footer-note" />
      <p>{postedAt(createdAt)}</p>
    </section>

  );
}

const userShapeComment = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  avatar: PropTypes.string.isRequired,
};

CommentItem.propTypes = {
  content: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  owner: PropTypes.shape(userShapeComment).isRequired,
};

export { userShapeComment };

export default CommentItem;
