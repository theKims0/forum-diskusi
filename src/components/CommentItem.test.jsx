/**
 * testing secenario commponent comment item
 * should display the content correctly
 * should display the owner information correctly
 * should display the createdAt date correctly
 * renders correctly
 */
import React from 'react';
import {
  describe, it, expect, vi,
} from 'vitest';
import { render, screen } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';
import CommentItem from './CommentItem';

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Mocking the postedAt function
vi.mock('../utils', () => ({
  default: (date) => `Posted at ${date}`,
}));

describe('CommentItem component', () => {
  const owner = {
    id: 'user123',
    name: 'John Doe',
    avatar: 'https://example.com/avatar.jpg',
  };

  const commentProps = {
    content: 'This is a comment',
    createdAt: '2024-06-12T12:00:00Z',
    owner,
  };

  it('should display the content correctly', () => {
    // Arrange
    render(<CommentItem {...commentProps} />);

    // Assert
    expect(screen.getByText('This is a comment')).toBeInTheDocument();
  });

  it('should display the owner information correctly', () => {
    // Arrange
    render(<CommentItem {...commentProps} />);

    // Assert
    expect(screen.getByText('@user123')).toBeInTheDocument();
    expect(screen.getByText('John Doe')).toBeInTheDocument();
    expect(screen.getByAltText(owner)).toBeInTheDocument();
  });

  it('should display the createdAt date correctly', () => {
    // Arrange
    render(<CommentItem {...commentProps} />);

    // Assert
    expect(screen.getByText('Posted at 2024-06-12T12:00:00Z')).toBeInTheDocument();
  });

  it('renders correctly', () => {
    // Arrange
    const { container } = render(<CommentItem {...commentProps} />);

    // Assert
    expect(container).toMatchSnapshot();
  });
});
