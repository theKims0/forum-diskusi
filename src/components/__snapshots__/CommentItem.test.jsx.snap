// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentItem component > renders correctly 1`] = `
<div>
  <section
    class="thread-reply"
  >
    <header
      class="footer-note"
    >
      <img
        alt="[object Object]"
        class="img-round"
        src="https://example.com/avatar.jpg"
      />
      <div
        class="user-info"
      >
        <p>
          @
          user123
        </p>
        <span>
          John Doe
        </span>
      </div>
    </header>
    <article>
      <p>
        This is a comment
      </p>
    </article>
    <footer
      class="footer-note"
    />
    <p>
      Posted at 2024-06-12T12:00:00Z
    </p>
  </section>
</div>
`;
